### Домашнее задание к лекции "Базовые понятия статистики"

Будем осуществлять работу с непростым [набором данных](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.csv) о состоянии здоровья лошадей, испытывающих кишечные колики. 

#### Задание 1. Базовое изучение

Изучить представленный набор данных на основе [описания его столбцов](https://raw.githubusercontent.com/obulygin/pyda_homeworks/master/statistics_basics/horse_data.names) и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [54]:
import pandas as pd

In [55]:
horses = pd.read_csv('horse_data.csv', encoding='utf-8')

In [56]:
# Операция (1-да, 2-нет)
# Возраст (1-взрослая, 2-молодая (меньше 6 мес)'
# ID больницы (может быть не уникальным)
# Температура (норма 37,8). Может повышаться при заболевании, может снижаться при шоковом состоянии лошади
# Пульс (норма для взрослых 30-40, 20-25 для атлетичных скакунов). Повышается при болевых состояниях и шоке
# Часота дыхания (норма 8-10). Низкая полезность показателя из-за волатильности
# Тем-ра конеч-ей (1-нормальная, 2-повышенная, 3-прохладные, 4-холодные). Переход с 3 на 4 свидует о шоке. 2 должна корелировать с Температурой
# Пер-ий пульс (1-нормальный, 2-повышенный, 3-пониженный, 4-отсутствует)
# Слизистые (1=норм.розовый, 2=ярко-розовый, 3=бледно-розовый, 4=бледно-синюшный, 5=ярко-красный, 6=темно-синюшный) 1,2 - норм, 3 возм. шок, 4-6 проблемы кровообращения
# Капиляры (1 = < 3 секунд, 2 = >= 3 секунды)
# Боль (1=тревога, без боли, 2=депрессия, 3=прерывистая легкая боль, 4=периодическая сильная боль, 5=постоянная сильная боль) Не дискретная величина
# Перистальтика (1=гипермобильность, 2=нормальный, 3=гипомотиль, 4=отсутствует)
# Вздутие (1=нет, 2=легкое, 3=умеренное, 4=тяжелое) Важный параметр. Тяжелые формы требуют операции
# Газы (1=нет, 2=легкий, 3=значительный)
# Рефлюкс (1 = нет, 2 = > 1 литра, 3 = < 1 литра) Выше значение - выше шанс непроходимости кишечника
# Рефлюкс PH (шкала от 0 до 14, где 7 нейтрально. Yормальные значения находятся в диапазоне от 3 до 4)
# Кал (1 = нормальный, 2 = увеличено, 3 = уменьшено, 4 = отсутствует)
# Живот (1=нормальный, 2=другое, 3=твердый кал в толстой кишке, 4=растянутый тонкий кишечник, 5=растянутый толстый кишечник) 3 - лечение, 4,5 - хир-ое поражение
# Эритроциты (норма 30-50). Повышение уровня свидетельствует об обезвоживании
# Белок (норма 6,0-7,5 г/дл)
# Брюш полость (1 = ясная 2 = мутная, 3 = серозно-кровавый) 1 - норма, 2 и 3 - проблемы кишечника
# Брюш полость белок. (гр/дл) Чем выше значение, тем выше шанс наличия проблем с кишечником
# Итог (1=жив, 2=умер, 3=был усыплен)
# Хир-ое поражение (1 = Да, 2 = Нет). Было ли поражение тканей. Определяется или при вскрытии или при операции
# Поражение место
# Поражение тип
# Поражение подтип
# Патологии (1=Да, 2=Нет)

In [57]:
horses.columns = ['Операция', 'Возраст', 'ID больницы', 'Температура', 'Пульс', 'Частота дыхания', 'Тем-ра конеч-ей', 'Пер-ий пульс',
                  'Слизистые', 'Капиляры', 'Боль', 'Перистальтика', 'Вздутие', 'Газы', 'Рефлюкс', 'Рефлюкс PH', 'Кал', 'Живот',
                  'Эритроциты', 'Белок', 'Брюш полость', 'Брюш полость белок', 'Итог', 'Хир-ое поражение', 'Поражение место', 
                  'Поражение тип', 'Поражение подтип', 'Патологии'
                 ]

In [58]:
horses = horses.apply(pd.to_numeric, errors='coerce')
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 28 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Операция            298 non-null    float64
 1   Возраст             299 non-null    int64  
 2   ID больницы         299 non-null    int64  
 3   Температура         239 non-null    float64
 4   Пульс               275 non-null    float64
 5   Частота дыхания     241 non-null    float64
 6   Тем-ра конеч-ей     243 non-null    float64
 7   Пер-ий пульс        230 non-null    float64
 8   Слизистые           253 non-null    float64
 9   Капиляры            267 non-null    float64
 10  Боль                244 non-null    float64
 11  Перистальтика       255 non-null    float64
 12  Вздутие             243 non-null    float64
 13  Газы                196 non-null    float64
 14  Рефлюкс             194 non-null    float64
 15  Рефлюкс PH          53 non-null     float64
 16  Кал     

In [59]:
horses.describe().transpose()

count          mean           std       min       25%  \
Операция            298.0  1.395973e+00  4.898813e-01       1.0       1.0   
Возраст             299.0  1.642140e+00  2.177300e+00       1.0       1.0   
ID больницы         299.0  1.087748e+06  1.532026e+06  518476.0  528904.0   
Температура         239.0  3.816653e+01  7.335084e-01      35.4      37.8   
Пульс               275.0  7.193455e+01  2.868052e+01      30.0      48.0   
Частота дыхания     241.0  3.042739e+01  1.767826e+01       8.0      18.0   
Тем-ра конеч-ей     243.0  2.345679e+00  1.046369e+00       1.0       1.0   
Пер-ий пульс        230.0  2.013043e+00  1.042672e+00       1.0       1.0   
Слизистые           253.0  2.853755e+00  1.620294e+00       1.0       1.0   
Капиляры            267.0  1.303371e+00  4.766225e-01       1.0       1.0   
Боль                244.0  2.942623e+00  1.303993e+00       1.0       2.0   
Перистальтика       255.0  2.913725e+00  9.762978e-01       1.0       3.0   
Вздутие             243.0  2.259259e+00  1.061472e+00       1.0       1.0   
Газы                196.0  1.755102e+00  6.494372e-01       1.0       1.0   
Рефлюкс             194.0  1.582474e+00  8.053590e-01       1.0       1.0   
Рефлюкс PH           53.0  4.707547e+00  1.982311e+00       1.0       3.0   
Кал                 197.0  2.756345e+00  1.254186e+00       1.0       1.0   
Живот               181.0  3.685083e+00  1.492667e+00       1.0       2.0   
Эритроциты          270.0  4.630000e+01  1.043838e+01      23.0      38.0   
Белок               266.0  2.451729e+01  2.750906e+01       3.3       6.5   
Брюш полость        135.0  2.037037e+00  8.049048e-01       1.0       1.0   
Брюш полость белок  102.0  3.019608e+00  1.968567e+00       0.1       2.0   
Итог                298.0  1.550336e+00  7.379674e-01       1.0       1.0   
Хир-ое поражение    299.0  1.361204e+00  4.811548e-01       1.0       1.0   
Поражение место     299.0  3.632321e+03  5.390357e+03       0.0    2111.5   
Поражение тип       299.0  9.052843e+01  6.506371e+02       0.0       0.0   
Поражение подтип    299.0  7.387960e+00  1.277498e+02       0.0       0.0   
Патологии           299.0  1.668896e+00  4.713992e-01       1.0       1.0   

                          50%       75%        max  
Операция                 1.00       2.0        2.0  
Возраст                  1.00       1.0        9.0  
ID больницы         530310.00  534736.0  5305629.0  
Температура             38.20      38.5       40.8  
Пульс                   64.00      88.0      184.0  
Частота дыхания         24.00      36.0       96.0  
Тем-ра конеч-ей          3.00       3.0        4.0  
Пер-ий пульс             1.50       3.0        4.0  
Слизистые                3.00       4.0        6.0  
Капиляры                 1.00       2.0        3.0  
Боль                     3.00       4.0        5.0  
Перистальтика            3.00       4.0        4.0  
Вздутие                  2.00       3.0        4.0  
Газы                     2.00       2.0        3.0  
Рефлюкс                  1.00       2.0        3.0  
Рефлюкс PH               5.00       6.5        7.5  
Кал                      3.00       4.0        4.0  
Живот                    4.00       5.0        5.0  
Эритроциты              45.00      52.0       75.0  
Белок                    7.50      57.0       89.0  
Брюш полость             2.00       3.0        3.0  
Брюш полость белок       2.25       3.9       10.1  
Итог                     1.00       2.0        3.0  
Хир-ое поражение         1.00       2.0        2.0  
Поражение место       2322.00    3209.0    41110.0  
Поражение тип            0.00       0.0     7111.0  
Поражение подтип         0.00       0.0     2209.0  
Патологии                2.00       2.0        2.0

In [60]:
# Добавим столбец оценивающий признаки заболевания и прогнозирующий по этим данным состояние лошади
def check_horse_conditions(row):
    """
    1 - норма
    2 - легкая форма заболевания
    3 - тяжелая форма заболевания
    4 - шок
    """
    if row['Температура'] > 38 or row['Тем-ра конеч-ей'] == 2 or row['Боль'] == 4 or row['Вздутие'] == 3 or row['Брюш полость'] == 2:
        return 2
    elif row['Температура'] > 39 or row['Боль'] == 5 or row['Вздутие'] == 4 or row['Брюш полость'] == 3:
        return 3
    elif row['Температура'] < 37.2 or row['Тем-ра конеч-ей'] == 4 or row['Слизистые'] == 3:
        return 4
    else:
        return 1
    
horses['Состояние'] = horses.apply(check_horse_conditions, axis=1)

In [61]:
# Выбираем 8 столбцов для дальнейшего анализа
horses = horses[['Возраст', 'Пульс', 'Частота дыхания', 'Температура', 'Пер-ий пульс', 'Тем-ра конеч-ей', 'Боль', 'Операция', 'Состояние']]

#### Описание:

- Датафрейм содержит пропуски тербующие обработки (все столбцы с количеством не нулевых значний меньше 299)
- Столбцы "Возраст", имеют ошибки в заполнении данными (внесены не допустимые значения), которые необходимо исправить.
- Столбцы "Пульс", "Частота дыхания" возможно имеют выбросы требующие дополнительного анализа

#### Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [62]:
def find_ejection(column):
    q1 = horses[column].quantile(0.25)
    q3 = horses[column].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    remove_outliers = horses[horses[column].between(lower_bound, upper_bound, inclusive=True)].sort_values(column)
    result = pd.concat([horses, remove_outliers]).drop_duplicates(keep=False)
    return result.dropna(subset=[column])

In [63]:
find_ejection('Пульс')

Возраст  Пульс  Частота дыхания  Температура  Пер-ий пульс  \
2          9  164.0             84.0         39.1           1.0   
40         9  150.0             72.0         39.0           NaN   
54         9  160.0             20.0         38.6           NaN   
254        9  184.0             84.0         38.8           NaN   
274        9  150.0             50.0         38.8           3.0   

     Тем-ра конеч-ей  Боль  Операция  Состояние  
2                4.0   2.0       1.0          2  
40               NaN   NaN       2.0          2  
54               3.0   3.0       1.0          2  
254              1.0   4.0       1.0          2  
274              1.0   5.0       1.0          2

In [64]:
# Повышенная температура корелирует с повышенной частотой дыхания, а также возрастом. Можно предположить, что данные верны и соответствуют молодым особям
# В датафрейме меняем значение возраста 9 на возраст 2, соотвественно описанию.
horses['Возраст'] = horses['Возраст'].replace(9, 2)

In [65]:
find_ejection('Частота дыхания')

Возраст  Пульс  Частота дыхания  Температура  Пер-ий пульс  \
2          2  164.0             84.0         39.1           1.0   
38         2  146.0             96.0         39.2           NaN   
40         2  150.0             72.0         39.0           NaN   
81         2  100.0             80.0         38.1           1.0   
83         1   60.0             80.0         37.8           3.0   
102        2  140.0             68.0         38.0           1.0   
105        1   52.0             96.0         38.3           3.0   
119        1   54.0             66.0         39.4           1.0   
124        1   42.0             68.0         38.0           1.0   
207        1   88.0             80.0         37.8           3.0   
228        2  120.0             70.0         38.5           NaN   
243        2  124.0             88.0         38.2           3.0   
254        2  184.0             84.0         38.8           NaN   
264        1   56.0             68.0          NaN           1.0   
294        1  120.0             70.0          NaN           NaN   

     Тем-ра конеч-ей  Боль  Операция  Состояние  
2                4.0   2.0       1.0          2  
38               NaN   NaN       1.0          2  
40               NaN   NaN       2.0          2  
81               3.0   3.0       1.0          2  
83               1.0   2.0       1.0          2  
102              1.0   3.0       1.0          1  
105              NaN   NaN       1.0          2  
119              1.0   2.0       1.0          2  
124              4.0   3.0       1.0          4  
207              3.0   NaN       1.0          2  
228              NaN   NaN       1.0          2  
243              1.0   2.0       1.0          2  
254              1.0   4.0       1.0          2  
264              3.0   3.0       2.0          1  
294              4.0   2.0       1.0          4

In [66]:
# Для молодых особей оставляем как есть, для взрослых меняем на среднее значение (стандартное отклонение не высокое).
breath_min = find_ejection('Частота дыхания')['Частота дыхания'].min()
breath_mean = horses['Частота дыхания'].mean()
breath_mean

30.42738589211618

In [67]:
horses.loc[(horses['Возраст'] == 1) & (horses['Частота дыхания'] >= breath_min), 'Частота дыхания'] = breath_mean

In [68]:
horses['Частота дыхания'].mean()

28.625918975224252

#### Задание 3. Работа с пропусками

Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

1. Столбец "Операция"

In [69]:
horses[(horses['Операция'] != 1) & (horses['Операция'] != 2)]

Возраст  Пульс  Частота дыхания  Температура  Пер-ий пульс  \
131        1   48.0             20.0         38.0           1.0   

     Тем-ра конеч-ей  Боль  Операция  Состояние  
131              3.0   4.0       NaN          2

In [70]:
# судя по коду поражения лошадь оперировали
horses.loc[(horses['Операция'] != 1) & (horses['Операция'] != 2), 'Операция'] = 1

2. Столбец "Температура", "Пульс"

In [71]:
horses.groupby('Состояние')['Температура'].mean()

Состояние
1    37.740000
2    38.351685
3    37.600000
4    37.363636
Name: Температура, dtype: float64

In [72]:
horses.groupby('Состояние')['Пульс'].mean()

Состояние
1    58.742857
2    73.703125
3    82.218750
4    59.000000
Name: Пульс, dtype: float64

In [73]:
horses['Температура'].fillna(horses.groupby('Состояние')['Температура'].transform('median'), inplace=True)
horses['Пульс'].fillna(horses.groupby('Состояние')['Пульс'].transform('median'), inplace=True)

3. Столбец "Частота дыхания"

In [74]:
# Т.к. согласно описанию показатель волатильный и не информативный, то заполним данные медианным значением
horses['Частота дыхания'].fillna(horses['Частота дыхания'].median(), inplace=True)

4. Столбец "Тем-ра конеч-ей"

In [75]:
def get_mode(series):
    counts = {}
    for el in series:
        if el not in counts:
            counts[el] = 1
        else:
            counts[el] += 1
    maxw = 0
    mode_ = None
    for key, value in counts.items():
        if maxw < value:
            maxw = value
            mode_ = key
    return mode_

In [76]:
horses['Тем-ра конеч-ей'].fillna(horses.groupby('Температура')['Тем-ра конеч-ей'].transform(get_mode), inplace=True)

In [77]:
horses[horses['Тем-ра конеч-ей'].isna()]

Возраст  Пульс  Частота дыхания  Температура  Пер-ий пульс  \
112        1   60.0             28.0         36.8           NaN   
140        1   42.0             30.0         36.0           NaN   

     Тем-ра конеч-ей  Боль  Операция  Состояние  
112              NaN   NaN       1.0          4  
140              NaN   NaN       2.0          4

In [78]:
# Для оставшихся значений введем значение соответствующее описанию столбцов (Температура тела пониженная, предположительно шок, значение столбца 4)
horses['Тем-ра конеч-ей'].fillna(4, inplace=True)

4. Столбец "Боль"

In [85]:
horses['Боль'].fillna(horses.groupby('Состояние')['Боль'].transform(get_mode), inplace=True)

In [86]:
horses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 299 entries, 0 to 298
Data columns (total 9 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Возраст          299 non-null    int64  
 1   Пульс            299 non-null    float64
 2   Частота дыхания  299 non-null    float64
 3   Температура      299 non-null    float64
 4   Пер-ий пульс     230 non-null    float64
 5   Тем-ра конеч-ей  299 non-null    float64
 6   Боль             299 non-null    float64
 7   Операция         299 non-null    float64
 8   Состояние        299 non-null    int64  
dtypes: float64(7), int64(2)
memory usage: 21.1 KB


In [87]:
horses.describe()

Возраст       Пульс  Частота дыхания  Температура  Пер-ий пульс  \
count  299.000000  299.000000       299.000000   299.000000    230.000000   
mean     1.080268   71.117057        27.728584    38.129766      2.013043   
std      0.272162   27.767546        13.368299     0.672993      1.042672   
min      1.000000   30.000000         8.000000    35.400000      1.000000   
25%      1.000000   48.000000        20.000000    37.750000      1.000000   
50%      1.000000   64.000000        24.000000    38.100000      1.500000   
75%      1.000000   88.000000        32.000000    38.500000      3.000000   
max      2.000000  184.000000        96.000000    40.800000      4.000000   

       Тем-ра конеч-ей        Боль    Операция   Состояние  
count       299.000000  299.000000  299.000000  299.000000  
mean          2.297659    2.926421    1.394649    2.083612  
std           1.059486    1.242942    0.489595    0.692505  
min           1.000000    1.000000    1.000000    1.000000  
25%           1.000000    2.000000    1.000000    2.000000  
50%           3.000000    3.000000    1.000000    2.000000  
75%           3.000000    4.000000    2.000000    2.000000  
max           4.000000    5.000000    2.000000    4.000000